modules/mapper-extras/src/main/java/org/elasticsearch/index/query/ScoreScriptUtils.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.mapper.VectorEncoderDecoder;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.index.mapper.VectorEncoderDecoder.sortSparseDimsValues;
+
+public class ScoreScriptUtils {
+
+    //**************FUNCTIONS FOR DENSE VECTORS
+
+    /**
+     * Calculate a dot product between two dense vectors
+     *
+     * @param queryVector the query vector parsed as {@code List<Number>} from json
+     * @param docVectorBR BytesRef representing encoded document vector
+     */
+    public static float dotProduct(List<Number> queryVector, BytesRef docVectorBR){
+        float[] docVector = VectorEncoderDecoder.decodeDenseVector(docVectorBR);
+        return intDotProduct(queryVector, docVector);
+    }
+
+    /**
+     * Calculate cosine similarity between two dense vectors
+     *
+     * CosineSimilarity is implemented as a class to use
+     * painless script caching to calculate queryVectorMagnitude
+     * only once per script execution for all documents.
+     * A user will call `cosineSimilarity(params.queryVector, doc['my_vector'].getValue())`
+     */
+    public static final class CosineSimilarity {
+        final float queryVectorMagnitude;
+        List<Number> queryVector;
+
+        // calculate queryVectorMagnitude once per query execution
+        public CosineSimilarity(List<Number> queryVector) {
+            this.queryVector = queryVector;
+            float floatValue;
+            float dotProduct = 0f;
+            for (Number value : queryVector) {
+                floatValue = value.floatValue();
+                dotProduct += floatValue * floatValue;
+            }
+            this.queryVectorMagnitude = (float) Math.sqrt(dotProduct);
+        }
+
+        public float cosineSimilarity(BytesRef docVectorBR) {
```

Comment:
I would make these methods return a double. We only support floats at index time because of space contraints, but this isn't a problem here.

Reply:
@jpountz  Thanks for the review, Adrien.  I will change this to `double`.  The main reason for `float` was that it is a document's score, and all other Scorers are returning floats.